cloudinit/cmd/query.py

```diff
@@ -162,17 +165,45 @@ def handle_args(name, args):
             return 0
         return 1
 
-    response = convert_jinja_instance_data(instance_data)
+    jinja_without_aliases = convert_jinja_instance_data(instance_data)
+    jinja_aliases = convert_jinja_instance_data(
+        instance_data, include_key_aliases=True
+    )
+    response = jinja_without_aliases
     if args.varname:
-        try:
-            for var in args.varname.split('.'):
-                response = response[var]
-        except KeyError:
-            LOG.error('Undefined instance-data key %s', args.varname)
-            return 1
+        walked_key_path = ""
+        for key_path_part in args.varname.split('.'):
+            try:
+                # Walk key path using complete aliases dict, yet response
+                # should only contain jinja_without_aliases
+                jinja_aliases = jinja_aliases[key_path_part]
+            except KeyError:
+                if walked_key_path:
+                    msg = "instance-data '{key_path}' has no '{leaf}'".format(
+                        leaf=key_path_part, key_path=walked_key_path
+                    )
+                else:
+                    msg = "Undefined instance-data key '{}'".format(
+                        args.varname
+                    )
+                LOG.error(msg)
+                return 1
+            if key_path_part in response:
+                response = response[key_path_part]
+            else:  # We are an underscore_delimited key alias
+                for key in response:
+                    if get_jinja_variable_alias(key) == key_path_part:
+                        response = response[key]
+                        break
+            if walked_key_path:
+                walked_key_path += "."
+            walked_key_path += key_path_part
```

Comment:
I think this section is hard to follow in a function that's already fairly long. It might help to break it down some or maybe even just add some more comments, but I don't think it needs to block merging.

Reply:
118 line long function... I've done worse, but could do a lot better. breaking this down into a couple of separate functions for readability.